### ISSUE TITLE
Need hook after queue terminates and all event handlers are cleaned up.

### ISSUE THREAD

**Issue by jingidy** [dfellis/queue-flow#8]
I'm using queue-flow with mocha, and I need to be able to invoke the callback after the queue is complete and global variable `eventHandlers` is all cleaned up. 

I can't find any methods or events for this. I could just be calling the wrong things, so please bear with me, this is what I'm doing:

``` javascript
var plate = {};
q([bacon])
  .node(setBaconOnPlate)
  .node(function (cb) { 
    expect(plate.getBacon()).to.exist;
    cb();
  }.on('empty', function () { callback(); });
```

Calling callback() on 'empty' results in callback() executing before the chain is complete. However, calling callback() on 'close' results in mocha checking for leaks before global `eventHandlers` is cleared, since the close event is emitted before the rest of close is executed. 

Is there a hook for this kind of situation? It seems like 'after close' is needed in this case.

**Comment by dfellis** [comment]
Hmm...

Well, `empty` can fire multiple times if an earlier step in the chain stalls and doesn't provide new values faster than your current step can process them, so its definitely not the one to use.

`close` fires _before_ the actual closing of the queue because it's possible to `return false` and stop queue-flow from actually closing the queue, but I'm not sure what you mean by a global variable called `eventHandlers`? The `eventHandlers` variable in the 0.5.x branch was local to the queue-flow `Q` object; queue-flow 0.6.x has eliminated it entirely and broken that logic off into a separate library. If it ever became a global variable, that was a mistake.

Some ideas to help you out here:
1. Have you tried using the latest version of queue-flow (0.6.7) instead of the one you're using? (0.5.x)
2. If you can't upgrade for whatever reason, here is a hack to get an "after close" event:

``` js
var plate = {};
q([bacon])
    .node(setBaconOnPlate)
    .node(function (cb) { 
        expect(plate.getBacon()).to.exist;
        cb();
    })
    .toArray(function() { callback(); });
```

Each step in the queue-flow chain is actually a separate queue-flow object, and all `reduce`-like methods (ones that only return a single value based on the entire set of values passed through them) only call their callback _after_ they've totally deconstructed themselves, including the 0.5.x branch (see [here](https://github.com/dfellis/queue-flow/blob/0.5.34/lib/queue-flow.js#L232)).

That's not ideal, but I don't think your issue exists in the 0.6.x branch and `.on('close', myFunc)` ought to work fine on that one.

**Comment by jingidy** [comment]
I'm actually using 0.6.7 already. I can provide you with a test program later.

Thanks for the quick response!

**Comment by dfellis** [comment]
Are you using `lib/queue-flow.min.js` by any chance?

I just noticed that my automatic browserify and minification code is apparently not working; the minified version is out-of-date.

**Comment by jingidy** [comment]
I don't think so. I'm using it as a node module and it seems like that's using queue-flow.js

**Comment by dfellis** [comment]
``` bash
damocles@moya:~/oss/queue-flow(master)$ git grep eventHandlers
damocles@moya:~/oss/queue-flow(master)$ 
damocles@moya:~/oss/async-cancelable-events(master)$ git grep eventHandlers
damocles@moya:~/oss/async-cancelable-events(master)$ 
```

It _was_ returning a response on `lib/queue-flow.min.js`, which is why I was asking about that.

Please let me know when you have an example. I just wrote [this gist](https://gist.github.com/dfellis/5667015) and here's the result:

``` bash
damocles@moya:~/oss/5667015(master)$ npm test
npm WARN package.json queue-flow-leak-test@0.0.0 No README.md file found!

> queue-flow-leak-test@0.0.0 test /Users/damocles/oss/5667015
> mocha queue-flow-leak.test.js


  ․

  ✔ 1 test complete (505ms)

```

**Comment by jingidy** [comment]
Here's [a gist](https://gist.github.com/jingidy/5667768) showing the errors in 0.6.7. FWIW, everything works with 0.6.9!

**Comment by dfellis** [comment]
Figured it out.

There _was_ a global leak in 0.6.7; take a look at [the comparison between 0.6.7 and 0.6.8](https://github.com/dfellis/queue-flow/compare/0.6.7...0.6.8). I thought I had included the jshint change in 0.6.7, but I hadn't, so the latest version of queue-flow wasn't actually published.

Sorry about that!
